File: .skills/openclaw-skills/skills/0xnyk/xint/lib/analytics.ts
import { mkdirSync, writeFileSync, existsSync } from "fs";
import { join } from "path";
import { loadTokens, getValidToken } from "./oauth";
import { getUserTimeline, type TweetWithEngagement } from "./timeline";
import { trackCost } from "./costs";
import { bold, cyan, yellow, dim, green, red } from "./format";

const SKILL_DIR = import.meta.dir;
const EXPORTS_DIR = join(SKILL_DIR, "..", "data", "exports");

function ensureExportsDir(): void {
  if (!existsSync(EXPORTS_DIR)) mkdirSync(EXPORTS_DIR, { recursive: true });
}

function fmtPct(n: number): string {
  return (n * 100).toFixed(2) + "%";
}

function fmtNum(n: number): string {
  if (n >= 1_000_000) return (n / 1_000_000).toFixed(1) + "M";
  if (n >= 1_000) return (n / 1_000).toFixed(1) + "K";
  return String(n);
}

function median(arr: number[]): number {
  if (arr.length === 0) return 0;
  const sorted = [...arr].sort((a, b) => a - b);
  const mid = Math.floor(sorted.length / 2);
  return sorted.length % 2 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

function renderOverview(tweets: TweetWithEngagement[]): string {
  const totalImpressions = tweets.reduce((s, t) => s + t.metrics.impressions, 0);
  const totalLikes = tweets.reduce((s, t) => s + t.metrics.likes, 0);
  const totalReplies = tweets.reduce((s, t) => s + t.metrics.replies, 0);
  const totalRetweets = tweets.reduce((s, t) => s + t.metrics.retweets, 0);
  const rates = tweets.map(t => t.engagement_rate).filter(r => r > 0);
  const avgRate = rates.length > 0 ? rates.reduce((s, r) => s + r, 0) / rates.length : 0;
  const medRate = median(rates);

  let out = bold("Overview") + "\n";
  out += `  Tweets analyzed:    ${cyan(String(tweets.length))}\n`;
  out += `  Total impressions:  ${cyan(fmtNum(totalImpressions))}\n`;
  out += `  Total likes:        ${cyan(fmtNum(totalLikes))}\n`;
  out += `  Total replies:      ${cyan(fmtNum(totalReplies))}\n`;
  out += `  Total retweets:     ${cyan(fmtNum(totalRetweets))}\n`;
  out += `  Avg engagement:     ${yellow(fmtPct(avgRate))}\n`;
  out += `  Median engagement:  ${yellow(fmtPct(medRate))}\n`;
  return out;
}

function renderTopPerformers(tweets: TweetWithEngagement[]): string {
  const sorted = [...tweets].sort((a, b) => b.engagement_rate - a.engagement_rate);
  const top = sorted.slice(0, 5);

  let out = bold("Top Performers") + "\n";
  for (let i = 0; i < top.length; i++) {
    const t = top[i];
    const preview = t.text.replace(/https:\/\/t\.co\/\S+/g, "").trim().slice(0, 80);
    out += `  ${cyan(String(i + 1))}. ${yellow(fmtPct(t.engagement_rate))} engagement\n`;
    out += `     ${dim(preview)}${preview.length >= 80 ? "..." : ""}\n`;
    out += `     ${fmtNum(t.metrics.impressions)} imp | ${fmtNum(t.metrics.likes)} likes | ${fmtNum(t.metrics.retweets)} RT\n`;
    out += `     ${dim(t.tweet_url)}\n`;
  }
  return out;
}

function renderContentBreakdown(tweets: TweetWithEngagement[]): string {
  let threads = 0, media = 0, singles = 0;
  let threadEng = 0, mediaEng = 0, singleEng = 0;

  for (const t of tweets) {
    const hasMedia = (t.urls || []).some(u => u.url?.includes("pic.twitter.com")) ||
                     t.text.includes("pic.twitter.com");
    if (t.conversation_id === t.id) {
      threads++;
      threadEng += t.engagement_rate;
    } else if (hasMedia) {
      media++;
      mediaEng += t.engagement_rate;
    } else {
      singles++;
      singleEng += t.engagement_rate;
    }
  }

  let out = bold("Content Breakdown") + "\n";
  out += `  Threads:  ${cyan(String(threads))} tweets`;
  if (threads > 0) out += `  (avg ${yellow(fmtPct(threadEng / threads))} engagement)`;
  out += "\n";
  out += `  Media:    ${cyan(String(media))} tweets`;
  if (media > 0) out += `  (avg ${yellow(fmtPct(mediaEng / media))} engagement)`;
  out += "\n";
  out += `  Singles:  ${cyan(String(singles))} tweets`;
  if (singles > 0) out += `  (avg ${yellow(fmtPct(singleEng / singles))} engagement)`;
  out += "\n";
  return out;
}

function renderEngagementTrends(tweets: TweetWithEngagement[]): string {
  const sorted = [...tweets].sort((a, b) => a.created_at.localeCompare(b.created_at));
  if (sorted.length < 2) return bold("Engagement Trends") + "\n  Not enough data\n";

  // Group by day
  const byDay = new Map<string, { rates: number[]; impressions: number }>();
  for (const t of sorted) {
    const day = t.created_at.slice(0, 10);
    const entry = byDay.get(day) || { rates: [], impressions: 0 };
    entry.rates.push(t.engagement_rate);
    entry.impressions += t.metrics.impressions;
    byDay.set(day, entry);
  }

  let out = bold("Engagement Trends") + "\n";
  for (const [day, data] of byDay) {
    const avg = data.rates.reduce((s, r) => s + r, 0) / data.rates.length;
    const bar = "█".repeat(Math.min(Math.round(avg * 500), 30));
    out += `  ${dim(day)} ${bar} ${yellow(fmtPct(avg))} (${data.rates.length} tweets, ${fmtNum(data.impressions)} imp)\n`;
  }
  return out;
}

function buildJsonOutput(tweets: TweetWithEngagement[]): object {
  const rates = tweets.map(t => t.engagement_rate).filter(r => r > 0);
  return {
    summary: {
      tweet_count: tweets.length,
      total_impressions: tweets.reduce((s, t) => s + t.metrics.impressions, 0),
      total_likes: tweets.reduce((s, t) => s + t.metrics.likes, 0),
      total_replies: tweets.reduce((s, t) => s + t.metrics.replies, 0),
      total_retweets: tweets.reduce((s, t) => s + t.metrics.retweets, 0),
      avg_engagement_rate: rates.length > 0 ? rates.reduce((s, r) => s + r, 0) / rates.length : 0,
      median_engagement_rate: median(rates),
    },
    top_performers: [...tweets]
      .sort((a, b) => b.engagement_rate - a.engagement_rate)
      .slice(0, 5)
      .map(t => ({
        id: t.id,
        text: t.text.slice(0, 200),
        engagement_rate: t.engagement_rate,
        metrics: t.metrics,
        tweet_url: t.tweet_url,
        created_at: t.created_at,
      })),
    tweets: tweets.map(t => ({
      id: t.id,
      text: t.text,
      engagement_rate: t.engagement_rate,
      metrics: t.metrics,
      tweet_url: t.tweet_url,
      created_at: t.created_at,
    })),
  };
}

export async function cmdAnalytics(args: string[]): Promise<void> {
  let since = "7d";
  let asJson = false;
  let save = false;
  let pages = 3;

  let i = 0;
  while (i < args.length) {
    const arg = args[i];
    switch (arg) {
      case "--since":
        since = args[++i] || "7d";
        break;
      case "--json":
        asJson = true;
        break;
      case "--save":
        save = true;
        break;
      case "--pages":
        pages = Math.min(parseInt(args[++i] || "3"), 10);
        break;
      case "--help":
      case "-h":
        printHelp();
        return;
      default:
        break;
    }
    i++;
  }

  let token: string;
  try {
    token = await getValidToken();
  } catch (e: any) {
    console.error("OAuth required for analytics. Run 'xint auth setup' first.");
    console.error(`Error: ${e.message}`);
    process.exit(1);
    return;
  }

  const tokens = loadTokens();
  if (!tokens?.user_id) {
    console.error("Could not determine user ID from stored tokens. Re-run 'xint auth setup'.");
    process.exit(1);
    return;
  }

  console.error(`Fetching your tweets (since ${since})...`);

  const tweets = await getUserTimeline(tokens.user_id, token, {
    since,
    pages,
    exclude: ["replies", "retweets"],
  });

  trackCost("analytics", "analytics", 1);

  if (tweets.length === 0) {
    console.error("No tweets found in the specified period.");
    return;
  }

  console.error(`Analyzing ${tweets.length} tweets...\n`);

  if (asJson) {
    const data = buildJsonOutput(tweets);
    console.log(JSON.stringify(data, null, 2));
  } else {
    console.log(renderOverview(tweets));
    console.log(renderTopPerformers(tweets));
    console.log(renderContentBreakdown(tweets));
    console.log(renderEngagementTrends(tweets));
  }

  if (save) {
    ensureExportsDir();
    const date = new Date().toISOString().slice(0, 10);
    const path = join(EXPORTS_DIR, `analytics-${date}.json`);
    writeFileSync(path, JSON.stringify(buildJsonOutput(tweets), null, 2));
    console.error(`Saved to ${path}`);
  }
}

function printHelp(): void {
  console.log(`
Usage: xint analytics [options]

Analyze your account's tweet performance and engagement metrics.

Options:
  --since <dur>    Time period (default: 7d). e.g. 1d, 7d, 30d
  --pages <N>      Pages of tweets to fetch (default: 3, max 10)
  --json           Output as JSON
  --save           Save results to data/exports/analytics-{date}.json

Examples:
  xint analytics                     # Last 7 days
  xint analytics --since 30d         # Last 30 days
  xint analytics --json              # JSON output
  xint analytics --save              # Save to file
`);
}

File: .skills/openclaw-skills/skills/0xnyk/xint-rs/src/commands/analytics.rs
use anyhow::Result;
use colored::Colorize;
use std::collections::HashMap;
use std::fs;

use crate::api::twitter;
use crate::auth::oauth;
use crate::cli::AnalyticsArgs;
use crate::client::XClient;
use crate::config::Config;
use crate::costs;
use crate::format;
use crate::models::Tweet;

pub async fn run(args: &AnalyticsArgs, config: &Config, client: &XClient) -> Result<()> {
    let client_id = config.require_client_id()?;
    let (access_token, tokens) =
        oauth::get_valid_token(client, &config.tokens_path(), client_id).await?;

    eprintln!("Fetching your tweets (since {})...", args.since);

    let tweets = twitter::get_user_timeline(
        client,
        &access_token,
        &tokens.user_id,
        100,
        5,
        true,
        true,
        Some(&args.since),
    )
    .await?;

    costs::track_cost(
        &config.costs_path(),
        "timeline",
        &format!("/2/users/{}/tweets", tokens.user_id),
        tweets.len() as u64,
    );

    if tweets.is_empty() {
        println!("No tweets found in the specified time window.");
        return Ok(());
    }

    let stats = compute_stats(&tweets);

    if args.json {
        format::print_json_pretty_filtered(&stats)?;
    } else {
        println!(
            "{}",
            format_dashboard(&stats, &tokens.username, &args.since)
        );
    }

    if args.save {
        let exports_dir = config.exports_dir();
        fs::create_dir_all(&exports_dir)?;
        let path = exports_dir.join(format!(
            "analytics-{}-{}.json",
            tokens.username,
            chrono::Utc::now().format("%Y%m%d-%H%M%S")
        ));
        let json = serde_json::to_string_pretty(&stats)?;
        fs::write(&path, &json)?;
        eprintln!("Saved to {}", path.display());
    }

    costs::track_cost(&config.costs_path(), "analytics", "analytics", 0);

    Ok(())
}

#[derive(serde::Serialize)]
struct AnalyticsStats {
    overview: Overview,
    top_performers: Vec<TweetSummary>,
    content_breakdown: ContentBreakdown,
    engagement_trends: Vec<DayStats>,
}

#[derive(serde::Serialize)]
struct Overview {
    total_tweets: usize,
    total_impressions: u64,
    total_likes: u64,
    total_retweets: u64,
    total_replies: u64,
    avg_engagement_rate: f64,
    avg_likes_per_tweet: f64,
    avg_impressions_per_tweet: f64,
}

#[derive(serde::Serialize)]
struct TweetSummary {
    id: String,
    text_preview: String,
    likes: u64,
    impressions: u64,
    engagement_rate: f64,
    url: String,
}

#[derive(serde::Serialize)]
struct ContentBreakdown {
    with_links: usize,
    with_media: usize,
    with_hashtags: usize,
    with_mentions: usize,
    plain_text: usize,
}

#[derive(serde::Serialize)]
struct DayStats {
    date: String,
    tweets: usize,
    impressions: u64,
    likes: u64,
    engagement_rate: f64,
}

fn engagement_rate(t: &Tweet) -> f64 {
    if t.metrics.impressions == 0 {
        return 0.0;
    }
    let engagements = t.metrics.likes + t.metrics.retweets + t.metrics.replies + t.metrics.quotes;
    engagements as f64 / t.metrics.impressions as f64 * 100.0
}

fn compute_stats(tweets: &[Tweet]) -> AnalyticsStats {
    let total_impressions: u64 = tweets.iter().map(|t| t.metrics.impressions).sum();
    let total_likes: u64 = tweets.iter().map(|t| t.metrics.likes).sum();
    let total_retweets: u64 = tweets.iter().map(|t| t.metrics.retweets).sum();
    let total_replies: u64 = tweets.iter().map(|t| t.metrics.replies).sum();
    let total_quotes: u64 = tweets.iter().map(|t| t.metrics.quotes).sum();
    let n = tweets.len() as f64;

    let avg_engagement_rate = if total_impressions > 0 {
        (total_likes + total_retweets + total_replies + total_quotes) as f64
            / total_impressions as f64
            * 100.0
    } else {
        0.0
    };

    let overview = Overview {
        total_tweets: tweets.len(),
        total_impressions,
        total_likes,
        total_retweets,
        total_replies,
        avg_engagement_rate: (avg_engagement_rate * 100.0).round() / 100.0,
        avg_likes_per_tweet: (total_likes as f64 / n * 10.0).round() / 10.0,
        avg_impressions_per_tweet: (total_impressions as f64 / n).round(),
    };

    // Top 5 by engagement rate
    let mut sorted: Vec<&Tweet> = tweets.iter().collect();
    sorted.sort_by(|a, b| {
        engagement_rate(b)
            .partial_cmp(&engagement_rate(a))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    let top_performers: Vec<TweetSummary> = sorted
        .iter()
        .take(5)
        .map(|t| {
            let preview = if t.text.len() > 80 {
                format!(
                    "{}...",
                    &t.text[..t
                        .text
                        .char_indices()
                        .nth(77)
                        .map(|(i, _)| i)
                        .unwrap_or(t.text.len())]
                )
            } else {
                t.text.clone()
            };
            TweetSummary {
                id: t.id.clone(),
                text_preview: preview,
                likes: t.metrics.likes,
                impressions: t.metrics.impressions,
                engagement_rate: (engagement_rate(t) * 100.0).round() / 100.0,
                url: t.tweet_url.clone(),
            }
        })
        .collect();

    // Content breakdown
    let with_links = tweets.iter().filter(|t| !t.urls.is_empty()).count();
    let with_hashtags = tweets.iter().filter(|t| !t.hashtags.is_empty()).count();
    let with_mentions = tweets.iter().filter(|t| !t.mentions.is_empty()).count();
    let with_media = tweets
        .iter()
        .filter(|t| t.text.contains("https://t.co/") && t.urls.is_empty())
        .count();
    let plain_text = tweets.len() - with_links - with_media;

    let content_breakdown = ContentBreakdown {
        with_links,
        with_media,
        with_hashtags,
        with_mentions,
        plain_text,
    };

    // Daily engagement trends
    let mut by_day: HashMap<String, Vec<&Tweet>> = HashMap::new();
    for t in tweets {
        let day = if t.created_at.len() >= 10 {
            t.created_at[..10].to_string()
        } else {
            "unknown".to_string()
        };
        by_day.entry(day).or_default().push(t);
    }
    let mut engagement_trends: Vec<DayStats> = by_day
        .into_iter()
        .map(|(date, day_tweets)| {
            let impressions: u64 = day_tweets.iter().map(|t| t.metrics.impressions).sum();
            let likes: u64 = day_tweets.iter().map(|t| t.metrics.likes).sum();
            let total_eng: u64 = day_tweets
                .iter()
                .map(|t| {
                    t.metrics.likes + t.metrics.retweets + t.metrics.replies + t.metrics.quotes
                })
                .sum();
            let er = if impressions > 0 {
                total_eng as f64 / impressions as f64 * 100.0
            } else {
                0.0
            };
            DayStats {
                date,
                tweets: day_tweets.len(),
                impressions,
                likes,
                engagement_rate: (er * 100.0).round() / 100.0,
            }
        })
        .collect();
    engagement_trends.sort_by(|a, b| b.date.cmp(&a.date));

    AnalyticsStats {
        overview,
        top_performers,
        content_breakdown,
        engagement_trends,
    }
}

fn compact_number(n: u64) -> String {
    if n >= 1_000_000 {
        format!("{:.1}M", n as f64 / 1_000_000.0)
    } else if n >= 1_000 {
        format!("{:.1}K", n as f64 / 1_000.0)
    } else {
        n.to_string()
    }
}

fn format_dashboard(stats: &AnalyticsStats, username: &str, since: &str) -> String {
    let o = &stats.overview;
    let mut out = String::new();

    out.push_str(&format!(
        "\n{}  @{} (last {})\n\n",
        "Account Analytics".bold(),
        username.cyan(),
        since
    ));

    // Overview
    out.push_str(&format!("  {} {}\n", "Overview".bold().underline(), ""));
    out.push_str(&format!(
        "  Tweets:       {}\n",
        o.total_tweets.to_string().yellow()
    ));
    out.push_str(&format!(
        "  Impressions:  {}\n",
        compact_number(o.total_impressions).yellow()
    ));
    out.push_str(&format!(
        "  Likes:        {}\n",
        compact_number(o.total_likes).yellow()
    ));
    out.push_str(&format!(
        "  Retweets:     {}\n",
        compact_number(o.total_retweets).yellow()
    ));
    out.push_str(&format!(
        "  Replies:      {}\n",
        compact_number(o.total_replies).yellow()
    ));
    out.push_str(&format!(
        "  Avg ER:       {}%\n",
        format!("{:.2}", o.avg_engagement_rate).green()
    ));
    out.push_str(&format!(
        "  Avg Likes:    {}/tweet\n",
        format!("{:.1}", o.avg_likes_per_tweet).yellow()
    ));

    // Top Performers
    if !stats.top_performers.is_empty() {
        out.push_str(&format!("\n  {}\n", "Top 5 Performers".bold().underline()));
        for (i, t) in stats.top_performers.iter().enumerate() {
            let preview = if t.text_preview.len() > 60 {
                format!("{}...", &t.text_preview[..57])
            } else {
                t.text_preview.clone()
            };
            out.push_str(&format!(
                "  {}. {} | {}% ER | {} likes | {} views\n     {}\n",
                i + 1,
                preview.dimmed(),
                format!("{:.2}", t.engagement_rate).green(),
                compact_number(t.likes),
                compact_number(t.impressions),
                t.url.dimmed()
            ));
        }
    }

    // Content Breakdown
    let cb = &stats.content_breakdown;
    out.push_str(&format!("\n  {}\n", "Content Breakdown".bold().underline()));
    out.push_str(&format!("  With links:    {}\n", cb.with_links));
    out.push_str(&format!("  With media:    {}\n", cb.with_media));
    out.push_str(&format!("  With hashtags: {}\n", cb.with_hashtags));
    out.push_str(&format!("  With mentions: {}\n", cb.with_mentions));
    out.push_str(&format!("  Plain text:    {}\n", cb.plain_text));

    // Daily Trends
    if !stats.engagement_trends.is_empty() {
        out.push_str(&format!("\n  {}\n", "Daily Trends".bold().underline()));
        out.push_str(&format!(
            "  {:<12} {:>6} {:>10} {:>8} {:>8}\n",
            "Date", "Tweets", "Impress.", "Likes", "ER%"
        ));
        for d in &stats.engagement_trends {
            out.push_str(&format!(
                "  {:<12} {:>6} {:>10} {:>8} {:>7}%\n",
                d.date,
                d.tweets,
                compact_number(d.impressions),
                compact_number(d.likes),
                format!("{:.2}", d.engagement_rate)
            ));
        }
    }

    out
}
